packages/workspace/src/browser/workspace-service.ts

```diff
@@ -102,30 +213,112 @@ export class WorkspaceService implements FrontendApplicationContribution {
         if (valid) {
             // The same window has to be preserved too (instead of opening a new one), if the workspace root is not yet available and we are setting it for the first time.
             // Option passed as parameter has the highest priority (for api developers), then the preference, then the default.
+            const rootToOpen = (await this.workspace).root;
             const { preserveWindow } = {
-                preserveWindow: this.preferences['workspace.preserveWindow'] || !(await this.root),
+                preserveWindow: this.preferences['workspace.preserveWindow'] || !(rootToOpen),
                 ...options
             };
-            await this.server.setRoot(rootUri);
+            await this.server.setWorkspace(rootUri);
             if (preserveWindow) {
                 this._root = valid;
             }
-            this.openWindow(uri, { preserveWindow });
+            await this.openWindow({ preserveWindow });
             return;
         }
         throw new Error(`Invalid workspace root URI. Expected an existing directory location. URI: ${rootUri}.`);
     }
 
     /**
-     * Clears current workspace root and reloads window.
+     * Adds a root folder to the workspace
+     * @param uri URI of the root folder being added
      */
-    close(): void {
-        this.doClose();
+    async addRoot(uri: URI): Promise<void> {
+        const workspaceRootChange = await this.workspace;
+        const workspaceFolder = workspaceRootChange.root;
+        if (!this.opened || !workspaceFolder) {
+            throw new Error('Folder cannot be added as there is no active folder in the current workspace.');
+        }
+
+        const rootToAdd = uri.toString();
+        const valid = await this.toValidRoot(rootToAdd);
+        if (valid) {
+            if (this._root && !this._workspaceDescription.roots.find(r => r.uri === valid.uri)) {
+                const configUri = await this.getWorkspaceConfigFileUri();
+                if (configUri) {
+                    if (!await this.fileSystem.exists(configUri.toString())) {
+                        await this.fileSystem.createFile(configUri.toString());
+                    }
+                    await this.writeRootFolderConfigFile(
+                        (await this.fileSystem.getFileStat(configUri.toString()))!,
+                        [...this._workspaceDescription.roots, valid]
+                    );
+                }
+            }
+            return;
+        }
+        throw new Error(`Invalid workspace root URI. Expected an existing directory location. URI: ${rootToAdd}.`);
+    }
+
+    /**
+     * Removes folder(s) from workspace.
+     */
+    async removeFolders(uris: URI[]): Promise<void> {
+        if (!this.opened) {
+            throw new Error('Folder cannot be removed as there is no active folder in the current workspace.');
+        }
+        const validUris: string[] = [];
```

Comment:
Does it really make sense to validate URIs to remove? We are not going to use them.